homeassistant/components/hassio/backup.py

```diff
@@ -0,0 +1,170 @@
+
+    async def async_download_backup(
+        self,
+        backup_id: str,
+        *,
+        path: Path,
+        **kwargs: Any,
+    ) -> None:
+        """Download a backup file."""
+        raise NotImplementedError
+
+    async def async_upload_backup(
+        self,
+        *,
+        path: Path,
+        metadata: BackupUploadMetadata,
+        **kwargs: Any,
+    ) -> None:
+        """Upload a backup."""
+        await get_supervisor_client(self._hass).backups.reload()
+
+    async def async_list_backups(self, **kwargs: Any) -> list[BaseBackup]:
+        """List backups."""
+        return [
+            BaseBackup(
+                backup_id=backup.slug,
+                date=backup.date.isoformat(),
+                name=backup.name,
+                protected=backup.protected,
+                size=backup.size,
+            )
+            for backup in await get_supervisor_client(self._hass).backups.list()
+        ]
+
+    async def async_get_backup(
+        self,
+        backup_id: str,
+        **kwargs: Any,
+    ) -> BaseBackup | None:
+        """Return a backup."""
+        backups = await self.async_list_backups()
+        for backup in backups:
+            if backup.backup_id == backup_id:
+                return backup
+        return None
+
+    def get_backup_path(self, backup_id: str) -> Path:
+        """Return the local path to a backup."""
+        return self._backup_dir / f"{backup_id}.tar"
+
+    async def async_delete_backup(self, backup_id: str, **kwargs: Any) -> None:
+        """Remove a backup."""
+        raise NotImplementedError
+
+
+class SupervisorBackupReaderWriter(BackupReaderWriter):
+    """Class for reading and writing backups in supervised installations."""
+
+    temp_backup_dir = Path("/cloud_backups")
+
+    def __init__(self, hass: HomeAssistant) -> None:
+        """Initialize the backup reader/writer."""
+        self._hass = hass
+
+    async def async_create_backup(
+        self,
+        *,
+        addons_included: list[str] | None,
+        agent_ids: list[str],
+        database_included: bool,
+        backup_name: str,
+        folders_included: list[str] | None,
+        on_progress: Callable[[BackupProgress], None] | None,
+        password: str | None,
+    ) -> tuple[NewBackup, asyncio.Task[tuple[BaseBackup, Path]]]:
+        """Create a backup."""
+        addons_included_set = set(addons_included) if addons_included else None
+        folders_included_set = set(folders_included) if folders_included else None
+
+        client = get_supervisor_client(self._hass)
+        backup = await client.backups.partial_backup(
+            supervisor_backups.PartialBackupOptions(
+                addons=addons_included_set,
+                folders=folders_included_set,  # type: ignore[arg-type]
+                homeassistant=True,
+                name=backup_name,
+                password=password,
+                compressed=True,
+                location=None,
```

Comment:
We should pick the first supervisor local agent here
Also, the supervisor API needs to be extended for creating a backup in /cloud_backups temp folder